lgc/builder/BuilderImpl.cpp

```diff
@@ -91,23 +90,25 @@ Value *BuilderImplBase::CreateIntegerDotProduct(Value *vector1, Value *vector2,
   Type *inputTy = vector1->getType();
   assert(inputTy->isVectorTy() && inputTy->getScalarType()->isIntegerTy());
   Value *scalar = nullptr;
+  Type *outputTy = accumulator->getType();
 
   const unsigned compBitWidth = inputTy->getScalarSizeInBits();
-  assert(compBitWidth == 8 || compBitWidth == 16);
-
-  bool isHwSupported = getPipelineState()->getTargetInfo().getGpuProperty().hasIntegerDot;
+  bool isHwSupported =
+      (getPipelineState()->getTargetInfo().getGpuProperty().hasIntegerDot && (compBitWidth == 8 || compBitWidth == 16));
   const bool isSigned = (flags & lgc::Builder::FirstVectorSigned);
   const bool isMixed = (flags == lgc::Builder::FirstVectorSigned);
   // The mixed mode "First vector is signed and secont vector is unsigned" is not supported
   if (isHwSupported && isMixed)
     isHwSupported = false;
 
-  const bool hasNoAccumulator = isa<ConstantInt>(accumulator) && cast<ConstantInt>(accumulator)->isNullValue();
+  const bool hasAccumulator = !(isa<ConstantInt>(accumulator) && cast<ConstantInt>(accumulator)->isNullValue());
   const unsigned compCount = cast<FixedVectorType>(inputTy)->getNumElements();
+  Type *targetTy = isHwSupported ? getInt32Ty() : getInt64Ty();
+  accumulator = isSigned ? CreateSExt(accumulator, targetTy) : CreateZExt(accumulator, targetTy);
   if (!isHwSupported) {
     // Emulate dot product with no HW support cases
-    scalar = getInt32(0);
-    Type *targetTy = getInt32Ty();
+    scalar = getInt64(0);
```

Comment:
Does this generate worse code in the case that the result type is i32 and the inputs are vectors of i8 or i16 with mixed mode that is not hw supported? Should we maybe have an i32 scalar with saturating i32 adds in that case?